cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java

```diff
@@ -70,6 +73,208 @@ public LineageOperationsProcessor(Set<Connection> stageConnections,
       stageOutputsWithOrigins.put(stage, new LinkedHashMap<>());
     }
     this.noMergeRequiredStages = noMergeRequiredStages;
+    validate(stageInputFields, stageOutputFields, stageOperations);
+  }
+
+  /**
+   * Validate the inputs and outputs of operations.
+   *
+   * For each operation the input must be coming from the input schema that stage receives
+   * or it must be one of the output of operations recorded by that stage prior to the
+   * current operation. For example, consider that stage gets input fields as [a, b, c].
+   * First operation recorded by the stage generates outputs [d, e]. For any subsequent
+   * operation, the valid input set will be [a, b, c, d, e].
+   *
+   * For each operation the output generated by it must be the part of output schema or
+   * used as an input by the subsequent operations recorded by the stage. For example,
+   * if operation OP1 creates outputs [a, b, c, x] and stage has output fields as [a, b, d, e],
+   * then output c must be consumed by any subsequent operation, otherwise validation
+   * fails. It is also possible to generate the redundant outputs by operations. For example,
+   * consider other operations OP2 with inputs as [c] and output as [x], and OP3 with input as
+   * [X] and output as [b]. In this case the output field [b] created by OP1 is redundant,
+   * since the field b of the output schema will come from OP3 and b is not used as input by any
+   * operation subsequent of OP1. Validation fails in this case.
+   *
+   * @param stageInputFields stage and corresponding set of input fields
+   * @param stageOutputFields stage and corresponding set of output fields
+   * @param stageOperations stage and corresponding List of field operations
+   * @throws IllegalArgumentException when the validation fails
+   */
+  @VisibleForTesting
+  void validate(Map<String, Set<String>> stageInputFields, Map<String, Set<String>> stageOutputFields,
+                Map<String, List<PipelineOperation>> stageOperations) {
+    // For a given stage, following map stores the name of the field to the List of operations
+    // for which the field appears in the output. Eventually this map will only contains the fields
+    // which are not used by any subsequent operation and are also not part of output schema of that stage.
+    Map<String, Map<String, List<String>>> unusedOutputsFromStage = new HashMap<>();
```

Comment:
nested collections like this are confusing. It looks like we don't actually require the outermost map, it's just being used to hold the stage name. It would be easier to understand if this were just a list of some new class that has the stage name as a field and the map of unused outputs as a field.